Implement popover integration with close watcher

lukewarlow opened the pull request:
#### 6b0e6bababc515e5035653c11780824ce168c377
<pre>
Implement popover integration with close watcher
<a href="https://bugs.webkit.org/show_bug.cgi?id=274719">https://bugs.webkit.org/show_bug.cgi?id=274719</a>

Reviewed by Anne van Kesteren.

Implement the internal close watcher for popovers.

Spec: <a href="https://html.spec.whatwg.org/multipage/popover.html#popover-close-watcher">https://html.spec.whatwg.org/multipage/popover.html#popover-close-watcher</a>

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/nnn-popovers-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/nyn-popovers-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-synthetic-keydown-expected.txt:
* Source/WebCore/dom/PopoverData.h:
(WebCore::PopoverData::closeWatcher):
(WebCore::PopoverData::setCloseWatcher):
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::PopoverData::PopoverCloseWatcherEventListener::PopoverCloseWatcherEventListener):
(WebCore::PopoverData::PopoverCloseWatcherEventListener::handleEvent):
(WebCore::HTMLElement::showPopoverInternal):
(WebCore::HTMLElement::hidePopoverInternal):
* Source/WebCore/html/closewatcher/CloseWatcher.cpp:
(WebCore::CloseWatcher::create):
* Source/WebCore/html/closewatcher/CloseWatcher.h:
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::defaultKeyboardEventHandler):

Canonical link: <a href="https://commits.webkit.org/313838@main">https://commits.webkit.org/313838@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1749c9cda10e00ca3976b3f0612aeb59386c2870

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/164302 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/37786 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/31357 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/173331 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/121554 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/166171 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/38120 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/37787 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/127654 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/89833 "layout-tests (failure)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/167261 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/29949 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/147685 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/108199 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/28996 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/27620 "Passed tests") | [✅ 🛠 gtk3-libwebrtc](https://ews-build.webkit.org/#/builders/173/builds/21095 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/138898 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/167/builds/25401 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/175857 "Built successfully") | 
| | [✅ 🛠 ios-safer-cpp](https://ews-build.webkit.org/#/builders/174/builds/28678 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/169/builds/27069 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/135965 "Passed tests") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/37457 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/31738 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/135934 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/36619 "Built successfully and passed tests") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/38243 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/147196 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/100598 "Built successfully") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/31247 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/170/builds/24052 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/37053 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/110097 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/36449 "Built successfully") | [✅ 🧪 mac-site-isolation](https://ews-build.webkit.org/#/builders/185/builds/2109 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/36699 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/36599 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/0fb261c5d42b55534be8c54d6bc965e38748d667. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/29117

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/7cb9fd1e46c7258d1c0f8502ed1e91bc550a4d58. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/29117

lukewarlow committed:
Implement popover integration with close watcher
https://bugs.webkit.org/show_bug.cgi?id=274719

Reviewed by NOBODY (OOPS!).

Implement the popover internal close watcher.

Spec: https://html.spec.whatwg.org/multipage/popover.html#popover-close-watcher

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/nnn-popovers-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/nyn-popovers-expected.txt:
* Source/WebCore/dom/PopoverData.h:
(WebCore::PopoverData::closeWatcher):
(WebCore::PopoverData::setCloseWatcher):
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::PopoverData::PopoverCloseWatcherEventListener::PopoverCloseWatcherEventListener):
(WebCore::PopoverData::PopoverCloseWatcherEventListener::handleEvent):
(WebCore::HTMLElement::showPopoverInternal):
(WebCore::HTMLElement::hidePopoverInternal):
* Source/WebCore/html/closewatcher/CloseWatcher.cpp:
(WebCore::CloseWatcher::create):
* Source/WebCore/html/closewatcher/CloseWatcher.h:
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::defaultKeyboardEventHandler):

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/d5e13e9a1a5b215e59a9e9ccfc2375064383fdb9. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/29117

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/fa1e1259feac3a00ed89f8d2167a9c189b146cba. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/29117

annevk reviewed (approved):
Modulo bots this looks good to me. Maybe @rniwa also wants to give it a quick glance.

lukewarlow committed:
Implement popover integration with close watcher
https://bugs.webkit.org/show_bug.cgi?id=274719

Reviewed by Anne van Kesteren.

Implement the internal close watcher for popovers.

Spec: https://html.spec.whatwg.org/multipage/popover.html#popover-close-watcher

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/nnn-popovers-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/nyn-popovers-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-synthetic-keydown-expected.txt:
* Source/WebCore/dom/PopoverData.h:
(WebCore::PopoverData::closeWatcher):
(WebCore::PopoverData::setCloseWatcher):
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::PopoverData::PopoverCloseWatcherEventListener::PopoverCloseWatcherEventListener):
(WebCore::PopoverData::PopoverCloseWatcherEventListener::handleEvent):
(WebCore::HTMLElement::showPopoverInternal):
(WebCore::HTMLElement::hidePopoverInternal):
* Source/WebCore/html/closewatcher/CloseWatcher.cpp:
(WebCore::CloseWatcher::create):
* Source/WebCore/html/closewatcher/CloseWatcher.h:
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::defaultKeyboardEventHandler):

Canonical link: https://commits.webkit.org/313838@main
